### PR TITLE
[Clang] [Sema] Ensure noexcept(typeid(E)) checks if E throws when needed

### PR DESCRIPTION
**Pull request by MitalAshok**
3ad31e12ccfc7db25f3cbedc4ee966e7099ac78f changed it so that not all potentially-evaluated `typeid`s were marked as potentially-throwing, but I forgot to check the subexpression if the null check of the `typeid` didn't potentially-throw. This adds that check.

### CODE REVIEW

**Comment by MitalAshok** [comment]
CC @Endilll @Sirraide

**Review by Sirraide — APPROVED** [review]
LGTM

File: clang/lib/Sema/SemaExceptionSpec.cpp [hunk 1]
```diff
@@ -1111,13 +1111,18 @@ static CanThrowResult canDynamicCastThrow(const CXXDynamicCastExpr *DC) {
 }
 
 static CanThrowResult canTypeidThrow(Sema &S, const CXXTypeidExpr *DC) {
-  if (DC->isTypeOperand())
+  // Operand is not evaluated, cannot possibly throw
```
Comment (tbaederr) [inline clang/lib/Sema/SemaExceptionSpec.cpp:1114]:
```suggestion
  // Operand is not evaluated, cannot possibly throw.
```

File: clang/lib/Sema/SemaExceptionSpec.cpp [hunk 2]
```diff
@@ -1111,13 +1111,18 @@ static CanThrowResult canDynamicCastThrow(const CXXDynamicCastExpr *DC) {
...
+  if (!DC->isPotentiallyEvaluated())
     return CT_Cannot;
 
   if (DC->isValueDependent())
     return CT_Dependent;
 
-  return DC->hasNullCheck() ? CT_Can : CT_Cannot;
+  // Can throw std::bad_typeid if a nullptr is dereferenced
```
Comment (tbaederr) [inline clang/lib/Sema/SemaExceptionSpec.cpp:1121]:
```suggestion
  // Can throw std::bad_typeid if a nullptr is dereferenced.
```

**Comment by Sirraide** [comment]
Do we already have tests for your most recent commit (i.e. dependent expressions)? Because I recall CI passing even when that wasn’t there, but that may have been my imagination.

**Comment by MitalAshok** [comment]
@Sirraide Yes, the tests did pass because there were no tests for this.

For reference:
 * `llvm:main` currently only handles these two wrong https://github.com/llvm/llvm-project/pull/95846/commits/89da8b3bcc678430fe4225c723e87914f2c378cd#diff-c35ff1c0977223e8ef2893fd21f3886c57b26346258850d8c879362fd9222381R113-R114 (https://godbolt.org/z/1j9Kc187f)
 * The first commit in this PR didn't handle `f1`, `f2` and `f3` properly, https://github.com/llvm/llvm-project/pull/95846/commits/77b5b6477f9af79777a35cfdf1fb4659e631be71#diff-c35ff1c0977223e8ef2893fd21f3886c57b26346258850d8c879362fd9222381R124-R141 (It treated them as `CT_Cannot` since it incorrectly assumed they weren't potentially evaluated because the dependent type isn't a polymorphic class type)
 * The third commit fixed that by moving the check afterwards

File: clang/test/SemaCXX/cxx0x-noexcept-expression.cpp [hunk 3]
```diff
@@ -115,3 +116,39 @@ static_assert(!noexcept(typeid((Polymorphic<true, false>&&) Polymorphic<true, fa
...
+
+template<bool B>
+struct X {
+  template<typename T> void f();
+};
+template<typename T>
+void f1() {
+  X<noexcept(typeid(*T{}))> dependent;
+  dependent.f<void>();
+  // expected-error@-1 {{use 'template' keyword to treat 'f' as a dependent template name}}
```
Comment (Sirraide) [inline clang/test/SemaCXX/cxx0x-noexcept-expression.cpp:128]: Perhaps I’m just being a bit slow today, but it took me a few seconds to realise that this is testing *whether* the `noexcept(...)` is correctly marked as dependent here; imo that deserves a comment so whoever looks at this next doesn’t end up as confused as I was just now

**Review by Sirraide — APPROVED** [review]
Alright, this does look like we’re handling all cases now (that I can think of at least).

However, I’d like to see some comments added to `f1`–`f3`, indicating why they’re dependent, because it took me a bit of thinking to figure that out just now...

**Review by Sirraide — APPROVED** [review]
Thanks, the comments are pretty much exactly what I had in mind.

**Comment by Sirraide** [comment]
In case anyone ends up being confused by the lack of a release note here: #95718 already added one that still covers everything done in this pr seeing as this is mostly a follow-up to that one.

**Pull request merged**